AirCasting/Utils/AveragingService.swift

```diff
@@ -0,0 +1,173 @@
+  * - for streams containing >9 hrs of data apply a 60-second avg. time interval
+  *
+  * Use case:
+  * - the mobile active session reaches a duration of 2 hours and 5 seconds.
+  * - at 2 hours and 5 seconds, the first 5-second average measurement is plotted on the map and graph.
+  * - all of the data from the prior 2 hours is transformed into 5-second averages and the map and graph and stats are updated accordingly.
+  *
+  * Notes:
+  * - averages should be attached to the middle value geocoordinates and timestamps,
+  * i.e. if its a 5-second avg spanning the time frame 10:00:00 to 10:00:05,
+  * the avg value gets pegged to the geocoordinates and timestamp from 10:00:03.
+  * - thresholds are calculated based on ellapsed time of session, not based on actual measurements records
+  * (pauses in sessions are not taken into account)
+  * - if there are any final, unaveraged measurements on 2h+ or 9h+ session which would not fall into full averaging window
+  * (5s or 60s) they should be deleted
+  * - on threshold crossing (at 2h and at 9h into session) we also trim measurements that not fit into given window size
+  *
+  **/
+
+//TODO:
+// 5. get proper averaging window (for threshold)
+
+struct AvgMeasurement {
+    let window: AveragingWindow = .zeroWindow
+    let treshold: TimeThreshold
+}
+
+enum AveragingWindow: Int {
+    case zeroWindow = 1
+    case firstThresholdWindow = 5
+    case secondThresholdWindow = 60
+}
+
+enum TimeThreshold: Int {
+    // Two hours: 60 * 60 * 2 = 7200
+    case firstThreshold = 7200
+    // Nine hours: 60 * 60 * 9 = 32400
+    case secondThreshold = 32400
+}
+
+final class AveragingService: NSObject, ObservableObject {
+    // TODO:
+    // 1. calculate average
+    // 2. remove unaveraged measurements
+    // 3. mark measurements as averaged
+    // 4. setup timer for new session
+    // 5. perform averaging of previous measurements after crossing threshold
+    // 5. get proper averaging window (for threshold)
+    
+    private var sessionEntity: SessionEntity?
+    private let measurementStreamStorage: MeasurementStreamStorage
+    private var timers: [SessionUUID : AnyCancellable] = [:]
+    private var fetchedResultsController: NSFetchedResultsController<SessionEntity>?
+    
+    init(measurementStreamStorage: MeasurementStreamStorage) {
+        self.measurementStreamStorage = measurementStreamStorage
+        super.init()
+        
+        measurementStreamStorage.accessStorage { [weak self] storage in
+            let request: NSFetchRequest<SessionEntity> = SessionEntity.fetchRequest()
+            request.predicate = NSPredicate(format: "type == %@ AND status == %i",
+                                            SessionType.mobile.rawValue,
+                                            SessionStatus.RECORDING.rawValue)
+            request.sortDescriptors = [NSSortDescriptor(key: "type", ascending: true)]
```

Comment:
If we are taking only mobile, do we need to sort by type?

Reply:
We just have to have a sortDescriptors defined, and in this case, it just doesn't matter what descriptor we use, so I picked random thing here.